ctru-rs/src/applets/swkbd.rs

```diff
@@ -311,6 +368,71 @@ impl SoftwareKeyboard {
         self.state.filter_flags = filters.bits();
     }
 
+    /// Configure a custom filtering function to validate the input.
+    ///
+    /// The callback function will return a callback result and the error message to display when the input is invalid.
+    ///
+    /// # Notes
+    ///
+    /// The filter callback will work only for the next time the keyboard is used. After using it once, it must be set again.
+    /// 
+    /// # Example
+    ///
+    /// ```
+    /// # let _runner = test_runner::GdbRunner::default();
+    /// # fn main() {
+    /// #
+    /// use ctru::applets::swkbd::{SoftwareKeyboard, CallbackInput};
+    /// let mut keyboard = SoftwareKeyboard::default();
+    /// 
+    /// keyboard.set_filter_callback(|text| {
+    ///     if text.contains("boo") {
+    ///         println!("Ah, you scared me!");
+    ///     }
+    /// 
+    ///     (CallbackResult::Ok, None)
+    /// });
+    /// #
+    /// # }
+    pub fn set_filter_callback<F>(&mut self, callback: F)
+    where
+        F: FnOnce(&str) -> (CallbackResult, Option<CString>),
+    {
+        unsafe extern "C" fn internal_callback<F>(
+            user: *mut libc::c_void,
+            pp_message: *mut *const libc::c_char,
+            text: *const libc::c_char,
+            _text_size: libc::size_t,
+        ) -> ctru_sys::SwkbdCallbackResult
+        where
+            F: FnOnce(&str) -> (CallbackResult, Option<CString>),
+        {
+            let closure = Box::from_raw(user as *mut Box<F>);
+
+            let text = CStr::from_ptr(text);
+            let text_slice: &str = text.to_str().unwrap();
+
+            let result = closure(text_slice);
+
+            if let Some(cstr) = result.1 {
+                *pp_message = cstr.as_ptr();
+                Box::leak(Box::new(cstr)); // Definitely SHOULD NOT do this, but as far as design goes, it's clean.
```

Comment:
There is an interesting issue with lifetimes here. Since the `libctru` code only asks back for a pointer to a string (the one that contains the "error message" to display to the user"), there is no check about the memory's validity over time time. This implies that, if the CString containing the message were to be allocated during the filter callback's runtime, the message text would be dropped and deallocated before being shown. Furthermore, being constrained by `extern "C"` FFI, we can't make any form of real lifetime checks ourselves.

As it stands, the `CString` is simply returned by the callback and then leaked, as I've deemed memory leaking a "suitable" drop-in solution for now, but alternatives are possible. I want to discuss about an elegant solution that doesn't disrupt the API too much.

Reply:
According to libctru, who is supposed to deallocate the string and when? We should try to emulate the memory management they expect.

Reply:
If you check the [source for swkbdSetFilterCallback](https://github.com/devkitPro/libctru/blob/master/libctru/source/applets/swkbd.c#L138-L143), it looks like libctru just stashes pointers to the CBs and expects the user to keep everything alive as long as it's needed.

And if that's the case, we should probably store that data in the `Swkbd` struct.

Reply:
I was trying to think of some clever ways to express that "the given closure must not borrow anything that lives shorter than `SoftwareKeyboard`", but couldn't come up with anything obvious after a few tries. One option might be something like this?

```rs
struct SoftwareKeyboard {
	// ...
	filter_result: Option<CString> // defaults to None
}

impl SoftwareKeyboard {
    pub fn set_filter_callback<F>(&mut self, callback: F)
    where
        F: FnOnce(&str, &mut Option<CString>) -> CallbackResult,
	{ 
		// probably requires an extra layer of indirection to capture self:
		let closure = |s| closure(self, s);
		// ...
		let result = closure(text_slice);
		// ...
	}
}
```

Then we'd always guarantee the filter result, if set, was alive as long as the keyboard was alive?

Reply:
I think something like this would work? I haven't really tested though:

```rs
// This can't be `Clone` anymore if we have a boxed closure in there...
pub struct SoftwareKeyboard {
    state: Box<SwkbdState>,
    callback: Option<Box<dyn Fn(&CStr, &mut Option<CString>) -> CallbackResult>>,
    filtered_string: Option<CString>,
}

impl SoftwareKeyboard {
    pub fn set_filter_callback<F>(&mut self, callback: F)
    where
        F: Fn(&CStr, &mut Option<CString>) -> CallbackResult + 'static,
    {
        extern "C" fn internal_callback(
            user: *mut libc::c_void,
            pp_message: *mut *const libc::c_char,
            text: *const libc::c_char,
            _text_size: libc::size_t,
        ) -> ctru_sys::SwkbdCallbackResult {
            let this: *mut SoftwareKeyboard = user.cast();

            unsafe {
                let text = CStr::from_ptr(text);
                let result = (*this).run_callback(text);

                if let Some(newstr) = &(*this).filtered_string {
                    *pp_message = newstr.as_ptr();
                }
                result.into()
            }
        }

        self.callback = Some(Box::new(callback));

        unsafe {
            ctru_sys::swkbdSetFilterCallback(
                self.state.as_mut(),
                Some(internal_callback),
                (self as *mut Self).cast(),
            );
        }
    }
}
```

Reply:
That would definitely work. Just note that what you called `filtered_string` isn't, in fact, the original message with "filters applied". It's actually just the error message shown to the user for why the string isn't valid.

Reply:
Ah, I misunderstood how the filter callback worked then – seems like it would be perfect for the example then

Reply:
Tell me if you prefer the current version, and what you think should be changed. I've also moved it to the general struct's `impl` block since it needs to be visible from `write_exact()`.